### PR TITLE
First delivery of FATs for mpJwt-1.2

### PR DESCRIPTION
**Pull request by c00crane**
Add FAT tests to validate the behavior or mpJwt-1.2 config options: header, cookie, algorithm and audience.
Test with the new attributes in/as:
server.xml
microprofile-config.properties in META-INF
microprofile-config properties under WEB-INF
environment variables
system properties
Also test that server.xml values over-ride others.

### CODE REVIEW

**Comment by c00crane** [comment]
#13571 
#13572 
#13573 
#13574

**Comment by c00crane** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat

**Comment by c00crane** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat

**Comment by c00crane** [comment]
Merged in key updates and runtime changes - trying build again!
#build
#spawn.fullfat.buckets=com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat

**Comment by c00crane** [comment]
Picked up more updates and kicked off build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_KqcGoP2iEeqZosoAJ-GPIQ

**Comment by c00crane** [comment]
#libbybot

**Comment by c00crane** [comment]
#libbybot 
#build
#spawn.fullfat.buckets=com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat,com.ibm.ws.security.jwtsso_fat

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+     *            - the expected behavior that we need to validate
+     * @throws Exception
+     */
+    public void genericConfigTest(LibertyServer rs_server, String builder, String location, String name, Expectations expectations) throws Exception {
+
+        String thisMethod = "genericConfigTest";
+        loggingUtils.printMethodName(thisMethod);
+
+        for (List<String> app : getTestAppArray(rs_server)) {
+            standardTestFlow(builder, app.get(0), app.get(1), location, name, expectations);
```
Comment (ayoho) [inline dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java:407]: Doing `app.get(0)` and `app.get(1)` to get the app root context and app (name?) is pretty unclear. Could the test app array actually be something like a list or set of classes with getters for the actual elements that make up the app/app string?

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+                                 Expectations expectations) throws Exception {
+        String testUrl = buildAppUrl(server, rootContext, theApp);
+        standardTestFlow(builder, testUrl, className, location, name, expectations);
+    }
+
+    public void standardTestFlow(String builder, String testUrl, String className, String location, String name) throws Exception {
+        standardTestFlow(builder, testUrl, className, location, name, null);
+    }
+
+    public void standardTestFlow(String builder, String testUrl, String className, String location, String name, Expectations expectations) throws Exception {
```
Comment (ayoho) [inline dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java:475]: Having a separate class kinda like we did with `TestSettings` would be useful for setting all of this information and avoiding nine overloaded `standardTestFlow` methods.

**Review by ayoho — APPROVED** [review]

File: dev/com.ibm.ws.security.mp.jwt.1.1_fat/fat/src/com/ibm/ws/security/mp/jwt11/fat/MPJwtConfigUsingBuilderTests.java [hunk 3]
```diff
@@ -411,7 +345,8 @@ public void MPJwtConfigUsingBuilderTests_authFilter_false() throws Exception {
     @Test
     public void MPJwtConfigUsingBuilderTests_Audience_Valid() throws Exception {
 
-        super.restoreTestServers();
+        // restore the server just in case a previous test changed the config (forcing a restore in tests that do NOT reconfig is quicker than having all tests restore when then finish)
+        resourceServer.restoreServerConfigurationAndWaitForApps();
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.1_fat/fat/src/com/ibm/ws/security/mp/jwt11/fat/MPJwtConfigUsingBuilderTests.java:349]: Just about every test in the class reconfigures the server when the test case starts.  There are very few tests that use the default config.  Skipping the restore at the end of each test (when the next test will reconfigure again) saves tons of time.  I'm saving 40-45 reconfigs in this one test class doing it this way.

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/.classpath [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="src" path="fat/src"/>
+	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="src" path="/fattest.simplicity"/>
+	<classpathentry kind="src" path="/com.ibm.ws.security.mp.jwt.1.1_fat"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.junit.JUNIT_CONTAINER/4"/>
+	<classpathentry kind="src" path="/com.ibm.ws.security.fat.common"/>
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/.classpath:10]: will remove

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/bnd.bnd [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		jwtSso-1.0,\
+		ldapRegistry-3.0,\
+		mpConfig-1.3,\
+		mpJwt-1.1,\
+		servlet-3.1,\
+		ssl-1.0,\
+		timedexit-1.0,\
+		jwtsso-1.0,\
+		federatedregistry-1.0,\
+		ldapregistry-3.0
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/bnd.bnd:34]: Removed - Copied them from the 1.1 fat when I started.

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/servers/com.ibm.ws.security.mp.jwt.1.2.fat/configs/rs_server_AltConfigInApp_badServerXmlHeader.xml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+							http://${fat.server.hostip}:${bvt.prop.security_2_HTTP_default}/jwt/defaultJWT,
+							https://${fat.server.hostname}:${bvt.prop.security_2_HTTP_default.secure}/jwt/defaultJWT,
+							https://${fat.server.hostip}:${bvt.prop.security_2_HTTP_default.secure}/jwt/defaultJWT,
+							https://localhost:${bvt.prop.security_2_HTTP_default}/oidc/endpoint/OidcConfigSample,
+							https://localhost:${bvt.prop.security_2_HTTP_default.secure}/oidc/endpoint/OidcConfigSample,
+							https://9.41.244.182:8947/jwt/defaultJWT,
+							testIssuer">
+		tokenHeader=“Authorization” 
+		cookieName=“Bearer” 
+		privateKey Name=“privateKeyAlias”
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/servers/com.ibm.ws.security.mp.jwt.1.2.fat/configs/rs_server_AltConfigInApp_badServerXmlHeader.xml:30]: was starting to create a new server.xml and didn't need it - file has been deleted

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/shared/config/mpConfigFeatures.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<server>
+
+	<featureManager>
+		<feature>timedexit-1.0</feature>
+		<feature>jsp-2.3</feature>
+		<feature>ssl-1.0</feature>
+		<feature>jwt-1.0</feature>
+		<feature>jaxrs-2.1</feature>
+		<feature>mpJwt-1.2</feature> 
+		<feature>mpConfig-1.4</feature>
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/shared/config/mpConfigFeatures.xml:10]: removed the extra features

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/shared/config/mp11ConfigFeatures.xml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+	<featureManager>
+		<feature>timedexit-1.0</feature>
+		<feature>jsp-2.3</feature>
+		<feature>servlet-3.1</feature>
+		<feature>ssl-1.0</feature>
+		<feature>jaxrs-2.0</feature>
+		<feature>jwt-1.0</feature>
+		<feature>mpJwt-1.1</feature> 
+		<feature>mpConfig-1.3</feature>
```
[the remark this answers is not in the document]
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/publish/shared/config/mp11ConfigFeatures.xml:11]: removed the features

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java (hunk 1 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java:407]: Added a testApp class with getters, constructor...

File: dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java (hunk 2 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.mp.jwt.1.2_fat/fat/src/com/ibm/ws/security/mp/jwt12/fat/sharedTests/MPJwt12MPConfigTests.java:475]: 1.1 and 1.2 have their own standardTestFlow methods (they really do have to do different things).
Have reduced the overloaded methods.

**Comment by c00crane** [comment]
#libbybot

**Comment by c00crane** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.mp.jwt.1.1_fat,com.ibm.ws.security.mp.jwt.1.2_fat,com.ibm.ws.security.jwtsso_fat

**Comment by c00crane** [comment]
After many code changes, and infrastructure issues, I finally have a good build...
Let's hope the merge works...

**Pull request merged**